### PR TITLE
FM2-155 : Update ObservationDAO and Service

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/FM2-155
This PR doesnot is the current one with new changes in regards to the current merged PR
 cc @ibacher

### CODE REVIEW

**Comment by sherrif10** [comment]
cc @ibacher  @corneliouzbett kindly review

**Comment by sherrif10** [comment]
Travis keep failing @corneliouzbett

**Comment by tendomart** [comment]
Did you say it is an older PR ?

Did you look at this Travis Faillure "api: Compilation failure: Compilation failure: 

[ERROR] /home/travis/build/openmrs/openmrs-module-fhir2/api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:[40,8] org.openmrs.module.fhir2.api.translators.impl.ObservationTranslatorImpl is not abstract and does not override abstract method toOpenmrsType(org.hl7.fhir.r4.model.Observation) in org.openmrs.module.fhir2.api.translators.ToOpenmrsTranslator

[ERROR] /home/travis/build/openmrs/openmrs-module-fhir2/api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:[39,8] org.openmrs.module.fhir2.api.impl.FhirObservationServiceImpl is not abstract and does not override abstract method getTranslator() in org.openmrs.module.fhir2.api.impl.BaseFhirService

??

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/ObservationTranslator.java [hunk 1]
```diff
@@ -53,4 +53,6 @@ default Obs toOpenmrsType(Obs existingObs, Observation observation) {
 	 * @return an updated version of the current obs
 	 */
 	Obs toOpenmrsType(Obs existingObs, Observation observation, Supplier<Obs> groupedObsFactory);
+	
+	Obs toOpenmrsType(Obs fhirObs);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/ObservationTranslator.java:57]: @sherrif10 The declaration here should be `Obs toOpenmrsType(Observation observation)`, i.e., translating from a FHIR Observation object to an OpenMRS Obs.

**Comment by sherrif10** [comment]
@tendomart  this is new PR , i closed the first one to make this current one because the former had merge conflicts

**Comment by sherrif10** [comment]
@ibacher hope its looks ok now, Anything else

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java [hunk 2]
```diff
@@ -31,10 +32,11 @@
 import org.springframework.stereotype.Component;
 import org.springframework.transaction.annotation.Transactional;
 
+@Getter(AccessLevel.PROTECTED)
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:35]: This should go right above the `@Setter` annotation.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java [hunk 3]
```diff
@@ -31,10 +32,11 @@
...
 @Component
 @Transactional
 @Setter(AccessLevel.PACKAGE)
-public class FhirObservationServiceImpl implements FhirObservationService {
+public class FhirObservationServiceImpl extends BaseFhirService<Observation, org.openmrs.Obs> implements FhirObservationService {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:39]: Remember that you need to rename the `observationTranslator` field to `translator`

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java (hunk 2 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:35]: O k let me do it now

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java (hunk 3 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:39]: Am not getting you by here renaming ObservationTranslator to translator

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:40]: The field called `observationTranslator` in this class should be named `translator`.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java [hunk 4]
```diff
@@ -43,7 +44,7 @@
 	private FhirObservationDao dao;
 	
 	@Autowired
-	private ObservationTranslator observationTranslator;
+	private Translator translator;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:47]: This should be `private ObservationTranslator translator`;

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java [hunk 5]
```diff
@@ -17,6 +17,7 @@
 import ca.uhn.fhir.rest.param.ReferenceParam;
 import ca.uhn.fhir.rest.param.StringAndListParam;
 import ca.uhn.fhir.rest.param.TokenAndListParam;
+import javassist.Translator;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:20]: Please remove this import.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java (hunk 4 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java:47]: What's happening here is that we're using Lombok to automatically generate a getter and setter for the field. By naming the field `translator`, Lombok automatically generates a `getTranslator()` function, which the `BaseFhirService` class requires.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java (hunk 5 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java]: Finished

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java (hunk 4 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java]: Finished

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImpl.java]: Actually i was asking my self why we change to translator but hanks

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java [hunk 6]
```diff
@@ -121,6 +121,11 @@ public Observation toFhirResource(Obs observation) {
 		return obs;
 	}
 	
+	@Override
+	public Obs toOpenmrsType(Observation fhirObservation) {
+		return new Obs();
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:126]: Last thing. This should be `return toOpenmrsType(new Obs(), fhirObservation);` That way, it actually does something.

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java]: Finished

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java]: This time travis has taken long to cope up with building

**Comment by sherrif10** [comment]
Did you see two commits , i have squashed them

**Comment by ibacher** [comment]
@sherrif10 I'm not too concerned about having only a single commit in the PR. We can always fix things when we merge it.

**Comment by sherrif10** [comment]
Haaaa , you are funny, so i think its ready for merge

**Comment by ibacher** [comment]
@sherrif10 Ah... so we have this error:

```
[ERROR] /home/travis/build/openmrs/openmrs-module-fhir2/api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirObservationServiceImplTest.java:[66,39] cannot find symbol

  symbol:   method setObservationTranslator(org.openmrs.module.fhir2.api.translators.ObservationTranslator)

  location: variable fhirObservationService of type org.openmrs.module.fhir2.api.impl.FhirObservationServiceImpl
```

So in the `FhirObservationServiceImplTest` you'll need to change the call to `setObservationTranslator(observationTranslator)` into `setTranslator(observationTranslator)`.

**Comment by sherrif10** [comment]
Let mr do that now

**Comment by sherrif10** [comment]
Was trying to correct the mistake

**Comment by ibacher** [comment]
Hi @sherrif10!

You've got a merge conflict. But this can be resolved without changing very much (the problem is I moved the location of a single function in the source code).

To do that, here are a series of git commands to run from the command line.

First, make sure your repo has a link to this repository (if you don't already–you can skip this step if you do).

```
git remote add upstream git@github.com:openmrs/openmrs-module-fhir2.git
```

This names the remote pointing to this repository `upstream`, which is the name I tend to use for this. If you already have this repository linked as a remote under a different name, just substitute that for where I have `upstream`.

Now we get the latest commits:

```
git fetch --prune upstream
```

Now we rebase these changes on top of the current master:

```
git rebase upstream/master
```

At this point, you'll have a conflict in the `ObservationFhirResourceProvider` class. Open this class in your IDE and find the lines that look like this:

```
<<<<<<< FM2-155 -- Incoming Change
	
	@History
	public List<Resource> getObservationHistoryById(@IdParam @NotNull IdType id) {
		Observation observation = observationService.get(id.getIdPart());
		if (observation == null) {
			throw new ResourceNotFoundException("Could not find Observation with Id " + id.getIdPart());
		}
		return observation.getContained();
	}
=======
>>>>>>> master -- Current Change
```

In this case, we can just delete them all. Then while still in that class, find the already existing `getObservationHistoryById()` and change the `getObservationByUuid()` method to `get()`. Commit that, and things will be all set.

**Comment by sherrif10** [comment]
Ok i know this was a bit hacky but following your steps, i got stuck in rebasing my master but i thik i havent broke anything

**Comment by sherrif10** [comment]
This method  getObservationByUuid(java.lang.String) is making things to fail, and it seems not to be present in ObservationFhirResourceProvider

**Comment by sherrif10** [comment]
Congratulations @ibacher  i have finally fought on about it, i think its ready for merge

**Comment by ibacher** [comment]
@sherrif10 Thanks for your effort on this!

**Comment by sherrif10** [comment]
Welcome @ibacher  , let me go ahead and do the personDAO and service

**Pull request merged**